ice: `'ty::Const::eval' called on a non-valtree-compatible type`

File: /home/matthias/vcs/github/glacier2/fixed/icemaker_omni/1405E6412BA4B2706BCE4F1D8CF3AD39F9BF72F26CEA1DE9618621EFF2DB97B1.rs

auto-reduced (treereduce-rust):
````rust
struct X<const FN: fn() = { || {} }>;

````

original:
````rust
struct X<const FN: fn() = {
    || {}
}>;

````

Version information
````
rustc 1.75.0-dev
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.75.0-dev
LLVM version: 17.0.2

````

Command:
`/home/matthias/.rustup/toolchains/local-debug-assertions/bin/rustc `

hm, with `2021-04-26-x86_64-unknown-linux-gnu`, this crashes like this:

```
error[E0658]: default values for const generic parameters are experimental
 --> icemaker.rs:1:25
  |
1 | struct X<const FN: fn() = { || {} }>;
  |                         ^^^^^^^^^^^
  |
  = note: see issue #44580 <https://github.com/rust-lang/rust/issues/44580> for more information
  = help: add `#![feature(const_generics_defaults)]` to the crate attributes to enable

error[E0601]: `main` function not found in crate `icemaker`
 --> icemaker.rs:1:1
  |
1 | struct X<const FN: fn() = { || {} }>;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ consider adding a `main` function to `icemaker.rs`

error: internal compiler error: compiler/rustc_middle/src/ty/context.rs:2681:21: No bound vars found for "expr || {} (hir_id=HirId { owner: DefId(0:3 ~ icemaker[317d]::X), local_id: 6 })" (HirId { owner: DefId(0:3 ~ icemaker[317d]::X), local_id: 6 })

thread 'rustc' panicked at 'Box<Any>', /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/std/src/panic.rs:59:5
stack backtrace:
   0:     0x7f2f20494480 - std::backtrace_rs::backtrace::libunwind::trace::hdcf4f90f85129e83
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x7f2f20494480 - std::backtrace_rs::backtrace::trace_unsynchronized::h2669e30cb82f6732
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f2f20494480 - std::sys_common::backtrace::_print_fmt::hfbda19e17f6db318
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x7f2f20494480 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h1a8751bf59281272
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x7f2f2050585f - core::fmt::write::h7aa6cd0067dca82a
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/core/src/fmt/mod.rs:1094:17
   5:     0x7f2f20488d85 - std::io::Write::write_fmt::hd7dd3a1df9b6befb
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/std/src/io/mod.rs:1584:15
   6:     0x7f2f204981db - std::sys_common::backtrace::_print::h551e9ec8a9fa8106
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x7f2f204981db - std::sys_common::backtrace::print::ha4b1c5e95fa040b3
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x7f2f204981db - std::panicking::default_hook::{{closure}}::h0b34c9ab7fb9f857
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/std/src/panicking.rs:208:50
   9:     0x7f2f20497cbd - std::panicking::default_hook::h3067e8318decd17a
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/std/src/panicking.rs:225:9
  10:     0x7f2f20cdd27d - rustc_driver::report_ice::h8bc35a5e54cf0d4e
  11:     0x7f2f204988f0 - std::panicking::rust_panic_with_hook::h81b8facc50f34daa
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/std/src/panicking.rs:595:17
  12:     0x7f2f21cabd5e - std::panicking::begin_panic::{{closure}}::hd7f35ccfec27cc18
  13:     0x7f2f21cabb39 - std::sys_common::backtrace::__rust_end_short_backtrace::h839d32355d5111c5
  14:     0x7f2f21cabcd2 - std::panicking::begin_panic::he6f44bfa2259e2cc
  15:     0x7f2f21ca0750 - std::panic::panic_any::ha59e5e1d194a13cd
  16:     0x7f2f21ca369a - rustc_errors::HandlerInner::bug::hfbe6aa453a0c7bd6
  17:     0x7f2f21ca3143 - rustc_errors::Handler::bug::hc3928b5f525c8142
  18:     0x7f2f21be8acc - rustc_middle::ty::context::tls::with_opt::h67d71acacbb2f7b2
  19:     0x7f2f21bedc30 - rustc_middle::util::bug::opt_span_bug_fmt::h2533ea0dc35d801c
  20:     0x7f2f21bedba6 - rustc_middle::util::bug::bug_fmt::h9409fc10762a69d4
  21:     0x7f2f226e106c - rustc_middle::ty::context::TyCtxt::late_bound_vars::h417f9d718f645d03
  22:     0x7f2f22a1042a - rustc_typeck::check::closure::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::supplied_sig_of_closure::h7bf901877c83d281
  23:     0x7f2f21ea1854 - rustc_typeck::check::closure::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_closure::h0b8771b352623475
  24:     0x7f2f21ea8c25 - rustc_typeck::check::expr::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_kind::h8ae8890e3d235cbd
  25:     0x7f2f21ea5885 - rustc_typeck::check::expr::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation::h7d9e36ee6fbf619f
  26:     0x7f2f21ebd41c - rustc_typeck::check::fn_ctxt::checks::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_block_with_expected::ha706ab2c822efecc
  27:     0x7f2f21ea5885 - rustc_typeck::check::expr::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation::h7d9e36ee6fbf619f
  28:     0x7f2f21f14921 - rustc_infer::infer::InferCtxtBuilder::enter::h6c7f9b49e9caad1b
  29:     0x7f2f21f7fe4d - rustc_typeck::check::typeck::he67f0c67dd42fb83
  30:     0x7f2f2210edb5 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::ha60de4e68c1c2b1f
  31:     0x7f2f220e9f02 - rustc_data_structures::stack::ensure_sufficient_stack::h73e95ffe8c981a72
  32:     0x7f2f220ba311 - rustc_query_system::query::plumbing::force_query_with_job::hf400d472addce63c
  33:     0x7f2f22076e14 - rustc_query_system::query::plumbing::get_query_impl::h0741a48aac2c95f2
  34:     0x7f2f22118b8d - <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::typeck::hfa88e0e6127efd00
  35:     0x7f2f226d846f - rustc_middle::ty::context::TyCtxt::typeck_opt_const_arg::ha110041f26dfaedb
  36:     0x7f2f2224c7b0 - rustc_mir::const_eval::eval_queries::eval_to_allocation_raw_provider::hd7eb305c219d6525
  37:     0x7f2f220cab1d - rustc_query_impl::<impl rustc_query_system::query::config::QueryAccessors<rustc_query_impl::plumbing::QueryCtxt> for rustc_query_impl::queries::eval_to_allocation_raw>::compute::h9366e8b03291c277
  38:     0x7f2f22110f76 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::hd2ef4fdb556b134e
  39:     0x7f2f220ec024 - rustc_data_structures::stack::ensure_sufficient_stack::hea35957a05928b00
  40:     0x7f2f220ad1cb - rustc_query_system::query::plumbing::force_query_with_job::h472bde5e03af862f
  41:     0x7f2f2209e347 - rustc_query_system::query::plumbing::get_query_impl::hd4e6d9e21a9393f5
  42:     0x7f2f22118dd7 - <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::eval_to_allocation_raw::h222a9162fae22499
  43:     0x7f2f2224b63b - rustc_mir::const_eval::eval_queries::eval_to_const_value_raw_provider::h7f8249e4f7a0cc11
  44:     0x7f2f220cab5d - rustc_query_impl::<impl rustc_query_system::query::config::QueryAccessors<rustc_query_impl::plumbing::QueryCtxt> for rustc_query_impl::queries::eval_to_const_value_raw>::compute::hd3ec2853454ad486
  45:     0x7f2f2210862a - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::h2aaf4628c58bc636
  46:     0x7f2f220eacc4 - rustc_data_structures::stack::ensure_sufficient_stack::hb8d43f9de8e7b018
  47:     0x7f2f220b2b77 - rustc_query_system::query::plumbing::force_query_with_job::h9c21c44dbbcdffcf
  48:     0x7f2f220966fb - rustc_query_system::query::plumbing::get_query_impl::hb528c52373b2c217
  49:     0x7f2f22118ef7 - <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::eval_to_const_value_raw::hdf5998513c48a889
  50:     0x7f2f226d3982 - rustc_middle::mir::interpret::queries::<impl rustc_middle::ty::context::TyCtxt>::const_eval_global_id::h7294e663d4978aae
  51:     0x7f2f226e9c9d - rustc_middle::mir::interpret::queries::<impl rustc_middle::ty::context::TyCtxt>::const_eval_resolve::h7fba165a21321c4f
  52:     0x7f2f23029d46 - rustc_infer::infer::InferCtxt::const_eval_resolve::hca3c298c3588be53
  53:     0x7f2f23021e23 - rustc_trait_selection::traits::const_evaluatable::is_const_evaluatable::h26e2a2545fabab56
  54:     0x7f2f225f5769 - rustc_trait_selection::traits::fulfill::FulfillProcessor::progress_changed_obligations::h49b2fd724ac87f00
  55:     0x7f2f22612022 - rustc_data_structures::obligation_forest::ObligationForest<O>::process_obligations::hb6a299a027ee1ef3
  56:     0x7f2f225f4d9b - <rustc_trait_selection::traits::fulfill::FulfillmentContext as rustc_infer::traits::engine::TraitEngine>::select_where_possible::h9a8565afc02fd26c
  57:     0x7f2f225f4bb0 - <rustc_trait_selection::traits::fulfill::FulfillmentContext as rustc_infer::traits::engine::TraitEngine>::select_all_or_error::h99db70751d1f457a
  58:     0x7f2f21eb4f4a - rustc_typeck::check::fn_ctxt::_impl::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::select_all_obligations_or_error::h090eec5fdbb06f7a
  59:     0x7f2f22a581e7 - rustc_infer::infer::InferCtxtBuilder::enter::ha8583eb4daf92417
  60:     0x7f2f21f988de - rustc_typeck::check::wfcheck::check_item_well_formed::h050c06be72496ecf
  61:     0x7f2f22110215 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::hc4b2fd7b87bf8ded
  62:     0x7f2f220aee7e - rustc_query_system::query::plumbing::force_query_with_job::h6a46056dcab7224b
  63:     0x7f2f2207eb6f - rustc_query_system::query::plumbing::get_query_impl::h372a567531ca4c6f
  64:     0x7f2f22119efd - <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::check_item_well_formed::hfb2c232624afeefb
  65:     0x7f2f21f9e0cc - <rustc_typeck::check::wfcheck::CheckTypeWellFormedVisitor as rustc_hir::intravisit::Visitor>::visit_item::h0b8817437244938d
  66:     0x7f2f21f8fcce - rustc_data_structures::sync::par_for_each_in::hc5ca15a054d22e20
  67:     0x7f2f22a628c9 - rustc_hir::hir::Crate::par_visit_all_item_likes::hf6e92d857a2b1ed6
  68:     0x7f2f22a32576 - rustc_session::session::Session::track_errors::h408406eec0de43c6
  69:     0x7f2f22a686e7 - rustc_typeck::check_crate::h94f0ca08944dc8ff
  70:     0x7f2f227ffc3d - rustc_interface::passes::analysis::hd5ac31040f53a69b
  71:     0x7f2f22c91785 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl::h2df1aabf2d2671a7
  72:     0x7f2f22c08857 - rustc_data_structures::stack::ensure_sufficient_stack::h219e3f3ae77528e1
  73:     0x7f2f22baa6d1 - rustc_query_system::query::plumbing::force_query_with_job::h091c05ae70f452e6
  74:     0x7f2f22b75b70 - rustc_query_system::query::plumbing::get_query_impl::h6dda0d4814804c64
  75:     0x7f2f22cadd3f - <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::analysis::h1f41c15d7c7a398d
  76:     0x7f2f227e033a - rustc_interface::passes::QueryContext::enter::h45323040f32afcbe
  77:     0x7f2f227c5295 - rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter::h724400ac5ffcdd64
  78:     0x7f2f227c4224 - rustc_span::with_source_map::h07c07459e999f3c9
  79:     0x7f2f227c6261 - rustc_interface::interface::create_compiler_and_run::h83442de2b4cadf7a
  80:     0x7f2f227df888 - scoped_tls::ScopedKey<T>::set::hb23c9cb7764fa643
  81:     0x7f2f227dfda3 - std::sys_common::backtrace::__rust_begin_short_backtrace::h81398fdbb964f101
  82:     0x7f2f227e35c5 - core::ops::function::FnOnce::call_once{{vtable.shim}}::hc030ad45e4e8f0df
  83:     0x7f2f204a7397 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h3aa31cb6360b59d9
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/alloc/src/boxed.rs:1546:9
  84:     0x7f2f204a7397 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h7719d3c7c5841461
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/alloc/src/boxed.rs:1546:9
  85:     0x7f2f204a7397 - std::sys::unix::thread::Thread::new::thread_start::hfbe13ead469fd0bc
                               at /rustc/3709ae324c9daabbdabe048faa6fbedbeccf27ec/library/std/src/sys/unix/thread.rs:71:17
  86:     0x7f2f202aa9eb - <unknown>
  87:     0x7f2f2032e7cc - <unknown>
  88:                0x0 - <unknown>

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.53.0-nightly (3709ae324 2021-04-25) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `X::{constant#0}`
#1 [eval_to_allocation_raw] const-evaluating + checking `X::{constant#0}`
#2 [eval_to_const_value_raw] simplifying constant for the type system `X::{constant#0}`
#3 [check_item_well_formed] checking that `X` is well-formed
#4 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0601, E0658.
For more information about an error, try `rustc --explain E0601`.
```

--- a/tests/ui/const-generics/default-ty-closure.rs
+++ b/tests/ui/const-generics/default-ty-closure.rs
@@ -1,0 +1,6 @@
+// https://github.com/rust-lang/rust/issues/116796
+
+struct X<const FN: fn() = { || {} }>;
+//~^ ERROR using function pointers as const generic parameters is forbidden
+
+fn main() {}

--- a/tests/ui/const-generics/default-ty-closure.stderr
+++ b/tests/ui/const-generics/default-ty-closure.stderr
@@ -1,0 +1,10 @@
+error: using function pointers as const generic parameters is forbidden
+  --> $DIR/default-ty-closure.rs:3:20
+   |
+LL | struct X<const FN: fn() = { || {} }>;
+   |                    ^^^^
+   |
+   = note: the only supported types are integers, `bool` and `char`
+
+error: aborting due to 1 previous error
+

